packages/astro/src/core/routing/manifest/create.ts

```diff
@@ -706,6 +714,49 @@ export function createRouteManifest(
 	};
 }
 
+async function getRoutePrerenderOption(
+	route: RouteData,
+	settings: AstroSettings,
+	fsMod: typeof nodeFs | undefined,
+	logger: Logger,
+) {
+	if (route.type !== 'page' && route.type !== 'endpoint') return;
+	const localFs = fsMod ?? nodeFs;
+	const content = await localFs.promises.readFile(
+		fileURLToPath(new URL(route.component, settings.config.root)),
+		'utf-8',
+	);
+
+	// Check if the route is pre-rendered or not
+	const match = /^\s*export\s+const\s+prerender\s*=\s*(true|false);?/m.exec(content);
+	if (match) {
+		route.prerender = match[1] === 'true';
+	}
+
+	// If not explicitly set, default to the global setting
+	if (typeof route.prerender === undefined) {
+		route.prerender = getPrerenderDefault(settings.config);
+	}
+
+	await runHookRouteSetup({ route, settings, logger });
```

Comment:
I think the global setting fallback should be after the `runHookRouteSetup` so that the hook could see that it's unset by the page.